### PR TITLE
[Modules] OPT1 - Update `nested_roleAssignment` to comply with `prefer-unquoted-property-names` linter rule

### PR DESCRIPTION
**Pull request by eriqua**
# Description

- Update `nested_roleAssignment` to comply with `prefer-unquoted-property-names` linter rule (Remove quotes from role definition names not containing spaces) 
- Update `nested_roleAssignment` with role definition list fetched by the Get-RoleAssignmentsList.ps1 utility
- Remove `prefer-unquoted-property-names` linter rule ignore from bicep.config
- Update Get-RoleAssignmentsList.ps1 utility
- Add Update-NestedRoleAssignment.ps1 utility
- Wiki update

> Note: unquoted role names are put on top for readability. Both categories (with and without quotes) are listed in alphabetical order

| Pipeline |
| - |
| [AnalysisServices: Servers](https://github.com/Azure/ResourceModules/actions/workflows/ms.analysisservices.servers.yml) |

# Type of Change

Please delete options that are not relevant.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Update to documentation

# Checklist

- [ ] I'm sure there are no other open Pull Requests for the same update/change
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (readme)
- [ ] I did format my code

### CODE REVIEW

File: utilities/tools/Update-NestedRoleAssignment.ps1 [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+Update all nested_roleassignments.bicep found in the library with latest available Role Definitions
+
+.EXAMPLE
+Update-NestedRoleAssignment -ProviderNamespace 'Microsoft.KeyVault' -ResourceType 'vaults'
+
+Update nested_roleassignments.bicep template for [Microsoft.KeyVault/vaults] module with latest available Role Definitions
+#>
+function Update-NestedRoleAssignment {
+
+    [CmdletBinding()]
```
Comment (MrMCake) [inline utilities/tools/Update-NestedRoleAssignment.ps1:101]:
```suggestion
    [CmdletBinding(SupportShouldProcess = $true)]
```

File: utilities/tools/Update-NestedRoleAssignment.ps1 [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+            Set-Location $modulesPath
+            $searchFile = Join-Path $modulesPath '**' $FileToSearch
+            $rbacPathList = Get-ChildItem -Path $searchFile -Recurse
+            foreach ($item in $rbacPathList) {
+                $FullFilePath = $item.FullName
+                $relativeFilePath = Get-Item $FullFilePath | Resolve-Path -Relative
+                $relativeFilePath = $relativeFilePath -replace '\.\\', ''
+                $relativeDirectoryPath = $relativeFilePath -replace '\\\.bicep\\nested_roleAssignments\.bicep', ''
+                $provider, $type = $relativeDirectoryPath -split '\\', 2
+                $null = Update-NestedRoleAssignmentInner -ProviderNamespace $provider -ResourceType $type -Verbose
```
Comment (MrMCake) [inline utilities/tools/Update-NestedRoleAssignment.ps1:136]:
```suggestion
                if ($PSCmdlet.ShouldProcess("Role Assignments for module [$relativeDirectoryPath]", 'Update')) {
                $null = Update-NestedRoleAssignmentInner -ProviderNamespace $provider -ResourceType $type -Verbose
                 }
```

File: utilities/tools/Update-NestedRoleAssignment.ps1 [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        $modulesPath = Join-Path $repoRootPath 'modules'
+        $FileToSearch = 'nested_roleAssignments.bicep'
+    }
+
+    process {
+        if ($ProviderNamespace -and $ResourceType) {
+            ########################################
+            ## Update RBAC roles for single module #
+            ########################################
+            $null = Update-NestedRoleAssignmentInner -ProviderNamespace $ProviderNamespace -ResourceType $ResourceType -Verbose
```
Comment (MrMCake) [inline utilities/tools/Update-NestedRoleAssignment.ps1:122]:
```suggestion
            if ($PSCmdlet.ShouldProcess("Role Assignments for module [$relativeDirectoryPath]", 'Update')) {
            $null = Update-NestedRoleAssignmentInner -ProviderNamespace $ProviderNamespace -ResourceType $ResourceType -Verbose
            }
```

File: utilities/tools/Update-NestedRoleAssignment.ps1 [hunk 4]
```diff
@@ -0,0 +1,148 @@
+﻿<#
```
Comment (MrMCake) [inline utilities/tools/Update-NestedRoleAssignment.ps1:1]: Suggestion: Should we call it `Update-NestedRoleAssignmentsMap/Object`?

File: utilities/tools/Get-RoleAssignmentsList.ps1 [hunk 5]
```diff
@@ -66,13 +66,22 @@ function Get-RoleAssignmentsList {
...
         foreach ($role in $relevantRoles | Sort-Object -Property 'Name' -Unique) {
-            $resBicep += "'{0}': subscriptionResourceId('Microsoft.Authorization/roleDefinitions','{1}')" -f $role.Name, $role.Id
+            if ($role.Name -match '\s') {
+                $resBicepQuote += "'{0}': subscriptionResourceId('Microsoft.Authorization/roleDefinitions','{1}')" -f $role.Name, $role.Id
+            } else {
+                $resBicepTop += "{0}: subscriptionResourceId('Microsoft.Authorization/roleDefinitions','{1}')" -f $role.Name, $role.Id
+            }
             $resArm += "`"{0}`": `"[subscriptionResourceId('Microsoft.Authorization/roleDefinitions','{1}')]`"," -f $role.Name, $role.Id
         }
+        $resBicep = $resBicepTop + $resBicepQuote
```
[the remark this answers is not in the document]
Comment (eriqua) [reply utilities/tools/Get-RoleAssignmentsList.ps1:82]: Fair point. I've created PR #2200 with the alternative approach to be discussed

[the remark this answers is not in the document]
Comment (MrMCake) [reply utilities/tools/Get-RoleAssignmentsList.ps1:82]: I guess automation is king? :D

[the remark this answers is not in the document]
Comment (eriqua) [reply utilities/tools/Get-RoleAssignmentsList.ps1:82]: Or queen 👑

**Pull request closed without merging**